### DISCUSSION TITLE
Spacy GraphQL interface

### DISCUSSION THREAD

**Discussion by oterrier** [explosion/spaCy#3568] [New Features & Project Ideas]
## Feature description
I have developped a tiny graphql wrapper aroung spacy using graphene and starlette.
I have found it very usefull to integrate spacy remotely with other programming languages like javascript or java/kotlin 

It is available in https://github.com/oterrier/gracyql

A kotlin client is available in https://github.com/oterrier/gracyql-kotlin

Maybe it could be interesting for the community ?

Cheers

Olivier Terrier

**Comment by ines** [comment]
Oh cool, thanks for sharing! If you'd like to submit a PR to the [spaCy Universe](https://spacy.io/universe) ([source here](https://github.com/explosion/spaCy/blob/master/website/meta/universe.json)), that'd be great.

I briefly experimented with spaCy and GraphQL [here](http://github.com/ines/spacy-graphql) (mostly to try out GraphQL and learn how it works) and it was really promising – your project seems like the more advanced version of that 😅

One thing I've been wondering: Do you know if it's possible to use the GraphQL wrapper to be smart about what pipeline components to run, depending on the query? For example, let's say the query only wants POS tags – instead of running the whole `nlp` pipeline (tagger, parser, NER, whatever, which can potentially be much slower), it could only run the components needed. For example:

```python
with nlp.disable_pipes(*[pipe for pipe in nlp.pipe_names if pipe != "tagger"]):
    doc = nlp(text)
    # and so on...
```

It seems like this is very much in line with the GraphQL philosophy – so if something like this could work with spaCy, that'd be a really useful feature.

**Comment by oterrier** [comment]
Hi Ines and thx for your comments
Sure, I'm going to submit a PR soon.

Regarding pipelines, it is already possible with my implementation:
```
query ParserDisabledQuery {
  nlp(model: "en", disable: ["parser", "ner"]) {
    doc(text: "I live in Grenoble, France") {
      text
      tokens {
        id
        pos
        lemma
        dep
      }
      ents {
        start
        end
        label
      }
    }
  }
}
```

**Comment by oterrier** [comment]
What is missing and could be cool is:
- automatic download of missing model if required
- a way to stream the result when many docs are provided in one shot (would require a kind of pagination or pubsub mecanism ...)

**Comment by ines** [comment]
Ahh, nice, I didn't see the `disabled` option!

I don't think this will be a big issue, but if you ever wanted to reduce the models kept in memory for different combinations of disabled pipes, using `nlp.disable_pipes` instead could be a good option. This way, you'd only ever have to load a model into memory once. And then on each request, you either run the full pipeline or disable components that aren't needed.

My initial idea was to do this automatically on the server and basically have a function that decides which components can be safely disabled, given the query. Could be nice for larger code bases – at any given point, you could just write a query for what you need, without having to think about what pipes to disable to make it as fast as possible.

> a way to stream the result when many docs are provided in one shot (would require a kind of pagination or pubsub mecanism ...)

That'd be super cool, yes! 👍

> automatic download of missing model if required

The only thing to consider here is: Would you actually want your server to quietly download and install a (potentially large) model in the background while fulfilling a request? But I guess it could be useful during development.

**Comment by oterrier** [comment]
Good idea, I've changed the way disable pipes is implemented to save memory.
Thx
BTW The PR for the Universe web site is issued
Olivier

**Comment by oterrier** [comment]
Hi Ines,
Just to let you know that I have implemented the batch mecanism with pagination as basically explained here https://github.com/oterrier/gracyql/blob/master/README.md#batch-multi-documents-query
Enjoy!

Olivier
